contrib/win32/win32compat/signal_wait.c

```diff
@@ -0,0 +1,220 @@
+*/
+
+#include "signal_internal.h"
+
+typedef struct _wait_for_multiple_objects_struct {
+	/* synchronization management */
+	HANDLE thread_handle;
+	HANDLE wait_event;
+
+	/* native function parameters input and output */
+	DWORD num_handles;
+	const HANDLE * handles;
+	DWORD return_value;
+}
+wait_for_multiple_objects_struct;
+
+DWORD WINAPI
+wait_for_multiple_objects_thread(LPVOID lpParam)
+{
+	wait_for_multiple_objects_struct *waitstruct = 
+		(wait_for_multiple_objects_struct *) lpParam;
+
+	/* wait for bin to complete -- this is alertable for our interrupt cleanup routine */
+	waitstruct->return_value = WaitForMultipleObjectsEx(waitstruct->num_handles,
+		waitstruct->handles, FALSE, INFINITE, TRUE);
+
+	/* notify the main thread that an event was found */
+	SetEvent(waitstruct->wait_event);
+
+	return TRUE;
+}
+
+VOID CALLBACK
+wait_for_multiple_objects_interrupter(_In_ ULONG_PTR dwParam)
+{
+	/* we must explicitly exit the thread since the thread could have been received
+	 * the alert prior to the thread running in which case it is acknowledged when
+	 * the threads starts running instead of when it is waiting at
+	 * WaitForMultipleObjectsEx */
+	ExitThread(0);
+}
+
+DWORD
+wait_for_multiple_objects_enhanced(_In_ DWORD  nCount, _In_ const HANDLE *lpHandles,
+	_In_ DWORD dwMilliseconds, _In_ BOOL bAlertable)
+{
+	/* if less than the normal maximum then just use the built-in function
+	 * to avoid the overhead of another thread */
+	if (nCount <= MAXIMUM_WAIT_OBJECTS) {
+
+		DWORD wait_ret = WaitForMultipleObjectsEx(nCount, lpHandles,
+			FALSE, dwMilliseconds, bAlertable);
+
+		if (wait_ret == WAIT_IO_COMPLETION) return WAIT_IO_COMPLETION_ENHANCED;
+		if (wait_ret == WAIT_TIMEOUT) return WAIT_TIMEOUT_ENHANCED;
+
+		/* translate normal offset to enhanced offset for abandoned threads */
+		if (wait_ret >= WAIT_ABANDONED_0 &&
+			wait_ret < WAIT_ABANDONED_0 + MAXIMUM_WAIT_OBJECTS) {
+			return WAIT_ABANDONED_0_ENHANCED + (wait_ret - WAIT_ABANDONED_0);
+		}
+
+		/* translate normal offset to enhanced offset for signaled threads */
+		if (wait_ret >= WAIT_OBJECT_0 &&
+			wait_ret < WAIT_OBJECT_0 + MAXIMUM_WAIT_OBJECTS) {
+			return WAIT_OBJECT_0_ENHANCED + (wait_ret - WAIT_OBJECT_0);
+		}
+
+		return WAIT_FAILED_ENHANCED;
+	}
+
+	/* number of separate bins / threads required to monitor execution */
+	const DWORD bin_size = MAXIMUM_WAIT_OBJECTS;
+	const DWORD bins_total = (nCount - 1) / bin_size + 1;
+
+	DWORD return_value = WAIT_FAILED_ENHANCED;
+
+	/* setup synchronization event to flag when the main thread should wake up */
+	HANDLE wait_event = CreateEvent(NULL, TRUE, FALSE, NULL);
+	if (wait_event == NULL) {
+		goto cleanup;
+	}
+
+	/* allocate an area to communicate with our threads */
+	wait_for_multiple_objects_struct *wait_bins = (wait_for_multiple_objects_struct *)
+		calloc(bins_total, sizeof(wait_for_multiple_objects_struct));
+	if (wait_bins == NULL) {
+		goto cleanup;
+	}
+
```

Comment:
ZeroMemory(wait_bins), otherwise, if anly thread creation fails below, you will be dealing with uninitialized memory in cleanup